oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,370 @@
+     - bundlephobia.com threw an error on the Paragon build, but we expect it’s the largest package here.
+   * - classnames
+     - 0.8k
+     -
+   * - core-js
+     - 241.1k
+     -
+   * - prop-types
+     - 0.9k
+     -
+   * - react
+     - 6.4k
+     -
+   * - react-dom
+     - 130.2k
+     -
+   * - react-redux
+     - 11.2k
+     -
+   * - react-router
+     - 58.9k
+     -
+   * - react-router-dom
+     - 77.1k
+     -
+   * - redux
+     - 3.7k
+     - Ultimately should be replaced with @reduxjs/toolkit. This document isn't making a judgment on the utility of redux, just noting that it is currently in use in the vast majority of MFEs.
+   * - regenerator-runtime
+     - 6.6k
+     -
+
+Total size: ~2,087.2k unzipped
+
+Sharing these dependencies has several important positive effects:
+
+- Enforces runtime consistency across all types of dependencies
+- Reduces runtime bundle size
+
+Note that "build time" and "dependency maintenance" are not mentioned above.
+
+Interestingly, because of the way Webpack module federation's dependency sharing is implemented, this may not have a large impact on build time or dependency maintenance. "Guest" MFEs still need to have the dependencies installed for the build to succeed, and artifacts for them will still be created and deployed - they just won't generally be used at runtime unless there's a version conflict.
+
+These dependencies still need to exist in the package.json file (and be installed by npm) for the build to succeed, and for IDE hinting, auto completion, linting, etc.
+
+Maintaining Dependency Consistency
+==================================
+
+This proposal fundamentally changes how we work with MFE dependencies, and will require us to adopt a more rigorous approach to ensuring dependency consistency and compatiblity.
+
+We expect that this may need to take a number of possible forms:
+
+Process
+-------
+
+We need to ensure maintainers and developers know what dependency versions to use, and when they need to upgrade to stay consistent. We also need a process to migrate through breaking changes in third-party dependencies.
+
+Best Practices
+--------------
+
+We need to ensure we minimize breaking changes in our own libraries (such as Paragon).
+
+Tooling
+-------
+
+Maintainers and developers should be warned of incompatibilities created by their PRs, or outside the repository by another project (such as the shell).
+
+Code Organization
+-----------------
+
+We could adopt a more drastic code organization scheme, such as a monorepo, which provides shared dependencies to core applications and libraries via a common package.json file, while otherwise maintaining applications' independent deployability.
+
+Alternately, we could create a new, versioned shared library which provides shared dependencies to MFE repositories as a single dependency (TBD how that would work).  This library could be paired with webpack aliases to prevent MFEs from needing to update their library import paths.
```

Comment:
We already do a little of this with `frontend-build`/`frontend-platform`, I could see having a `frontend-kitchensink` package as the *only* dependency in the `package.json` in every MFE as a (perhaps a bit overzealous) way to enforce dependency consistency between MFEs.

Reply:
> allow each MFE to select which major version of frontend-platform it uses

The issue there would come from [singletons](https://webpack.js.org/plugins/module-federation-plugin#singleton)

> #### `singleton`
> `boolean`
>
> This hint only allows a single version of the shared module in the shared scope (disabled by default). Some libraries use a global internal state (e.g. react, react-dom). Thus, it is critical to have only one instance of the library running at a time.

So if `react` made it into `frontend-platform` we'd need to *require* all MFEs use the same major version of `frontend-platform` for module federation to work.

Maybe it'd make sense to have 2 sets for each major release - the "this needs to be the same as the rest of the MFEs" set, and the "this can be upgraded independently" set?

Reply:
My understanding was that [as of React 17 you could have multiple versions of React in the same app](https://legacy.reactjs.org/blog/2020/10/20/react-v17.html#gradual-upgrades). Of course it's inefficient, but it's possible.

And in any case, React 17 and 18 are so similar and compatible that I don't anticipate that being a huge barrier; upgrading React 17 to 18 is usually simple. My concern would be more with Paragon (and in turn bootstrap) which relies on global [S]CSS rules (and maybe global JS too?). I don't think Paragon uses scoped CSS classes, so although you can have several different versions rendering your components, you have to pick which version gets to provide the CSS for the whole app - and then some of the components rendered by other versions may be broken. But this is challenging because (A) Paragon still needs a lot of development, and (B) when Paragon does introduce breaking changes it can require a lot of work to update the MFEs, at least in my limited recent experience.

Reply:
I've written up a description in the OEP of what I think a shared dependency library would look like.  It's a bit of conjecture on what the exact implementation would be, but I think it's worth exploring as an important component of the OEP.  Happy to talk about it more, and there may be an easy solution I'm not thinking of.